Make wrapPropagateTLSState more generic

lw opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #57638 Reduce overhead when Future invokes callbacks inline
* #57637 Introduce thenAsync method on Future
* #57636 Migrate from shared_ptr to intrusive_ptr for Future
* #57635 Pass reference to parent future in callbacks
* **#57634 Make wrapPropagateTLSState more generic**

`wrapPropagateTLSState` was restricting its argument to be an argument-less function, and I need to relax this for later work.

Also, it was requiring its argument to be converted to `std::function`, and also returned a `std::function`. Each creation of a `std::function` could cause a heap allocation. It's not particularly expensive, but here we can easily avoid it by having `wrapPropagateTLSState` directly operate on generic callables (thus, possibly, raw lambdas).

Differential Revision: [D28178782](https://our.internmc.facebook.com/intern/diff/D28178782/)

lw committed:
Make wrapPropagateTLSState more generic

`wrapPropagateTLSState` was restricting its argument to be an argument-less function, and I need to relax this for later work.

Also, it was requiring its argument to be converted to `std::function`, and also returned a `std::function`. Each creation of a `std::function` could cause a heap allocation. It's not particularly expensive, but here we can easily avoid it by having `wrapPropagateTLSState` directly operate on generic callables (thus, possibly, raw lambdas).

Differential Revision: [D28178782](https://our.internmc.facebook.com/intern/diff/D28178782/)

[ghstack-poisoned]

mrshenli commented on aten/src/ATen/ThreadLocalState.h:
```diff
@@ -87,12 +87,12 @@ class TORCH_API ThreadLocalStateGuard {
-std::function<T(void)> wrapPropagateTLSState(
-    std::function<T(void)> callback) {
-  return [tls_state = ThreadLocalState(), callback = std::move(callback)]() {
+auto wrapPropagateTLSState(T callback) {
```
Would I be correct if I assume we use `auto` here to avoid passing return type to the template?

lw replied:
We need `auto` here for the return type because the return value is a lambda, which has an "anonymous" type. In fact, the use of `auto` here is almost identical to the use of `auto` in expressions such as:
```
auto cb = [&](Foo f) { bar(f); };
```

Remember that lambdas, in fact, are just "syntactic sugar" and that under the hood a lambda defines a class (with a random unique name, in a way) in the local scope. The fields of this class are the objects that are captured in the lambda's closure, and the `operator()` of this class is the body of the lambda. Thus `auto` will correspond to this class. Because the class has no public name, we cannot refer to it directly, and `auto` or `decltype` are our only options.

We could use `std::function` instead of `auto` but this is _not_ equivalent, as it would mean that instead of returning the raw lambda as-is, we'd wrap it in some wrapper, possibly allocating heap memory, and moving data around, etc.

mrshenli reviewed (approved):
Nice simplification! Thanks! Test errors are irrelevant

rohan-varma commented on aten/src/ATen/ThreadLocalState.h:
```diff
@@ -87,12 +87,12 @@ class TORCH_API ThreadLocalStateGuard {
-  return [tls_state = ThreadLocalState(), callback = std::move(callback)]() {
+auto wrapPropagateTLSState(T callback) {
+  return [tls_state = ThreadLocalState(),
+          callback = std::move(callback)](auto&&... args) {
```
Thanks for this change! When I added this in https://github.com/pytorch/pytorch/pull/40326 I spent some amount of time trying to make it take variadic args but couldn't quite figure out how the argument forwarding works. I decided not to pursue it further at the time since we didn't really have multiple-arg callbacks in C++ future.

lw committed:
Update on "Make wrapPropagateTLSState more generic"


`wrapPropagateTLSState` was restricting its argument to be an argument-less function, and I need to relax this for later work.

Also, it was requiring its argument to be converted to `std::function`, and also returned a `std::function`. Each creation of a `std::function` could cause a heap allocation. It's not particularly expensive, but here we can easily avoid it by having `wrapPropagateTLSState` directly operate on generic callables (thus, possibly, raw lambdas).

Differential Revision: [D28178782](https://our.internmc.facebook.com/intern/diff/D28178782/)

[ghstack-poisoned]
